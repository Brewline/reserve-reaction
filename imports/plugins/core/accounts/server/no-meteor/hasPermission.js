const GLOBAL_GROUP = "__global_roles__";

/**
 * @name hasPermission
 * @method
 * @memberof Accounts
 * @param {Object} user - The user object, with `roles` property, to check.
 * @param {String[]} permissions - Array of permission strings. The account must have at least one of them either globally or for the roleGroup.
 * @param {String} [roleGroup] - The shop ID for which the permissions are needed, or a more specific roles group. If not set,
 *   only global roles will be checked.
 * @return {Boolean} True if the account with ID accountId has at least one of the requested permissions in the roleGroup group
 */
export default function hasPermission(user, permissions, roleGroup) {
  if (!user || !user.roles) return false;

  if (!Array.isArray(permissions)) throw new Error("permissions must be an array of strings");
  if (roleGroup !== undefined && roleGroup !== null && (typeof roleGroup !== "string" || roleGroup.length === 0)) {
    throw new Error("roleGroup must be a non-empty string");
  }

  const checkRoles = permissions.slice(0);

  // This should always return true for owners
  if (checkRoles.indexOf("owner") === -1) checkRoles.push("owner");

  const { roles } = user;

  // always check GLOBAL_GROUP
  const globalRoles = roles[GLOBAL_GROUP];
  if (Array.isArray(globalRoles) && checkRoles.some((role) => globalRoles.indexOf(role) !== -1)) return true;

  if (roleGroup) {
    // convert any periods to underscores for MongoDB compatibility
    const group = roleGroup.replace(/\./g, "_");

    const groupRoles = roles[group];
    if (Array.isArray(groupRoles) && checkRoles.some((role) => groupRoles.indexOf(role) !== -1)) return true;
  }

  return false;
}
