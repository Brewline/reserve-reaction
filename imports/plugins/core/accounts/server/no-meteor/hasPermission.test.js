import hasPermission from "./hasPermission";

test("returns false if no user", () => {
  const result = hasPermission(null, []);
  expect(result).toBe(false);
});

test("returns false if no user.roles", () => {
  const result = hasPermission({}, []);
  expect(result).toBe(false);
});

test("throws if permissions is not an array", () => {
  expect(() => hasPermission({ roles: {} })).toThrowErrorMatchingSnapshot();
});

test("throws if roleGroup is present but not a string", () => {
  expect(() => hasPermission({ roles: {} }, [], 123)).toThrowErrorMatchingSnapshot();
});

test("throws if roleGroup is present but an empty string", () => {
  expect(() => hasPermission({ roles: {} }, [], "")).toThrowErrorMatchingSnapshot();
});

test("returns true if in global role, even if not in group-scope role", () => {
  const result = hasPermission({
    roles: {
      __global_roles__: ["can_fry_bacon"], // eslint-disable-line camelcase
      scope: ["can_eat"]
    }
  }, ["can_fry_bacon", "can_scoop_ice_cream"], "scope");
  expect(result).toBe(true);
});

test("returns true if in group-scope role but not in global role", () => {
  const result = hasPermission({
    roles: {
      __global_roles__: ["can_fry_bacon"], // eslint-disable-line camelcase
      scope: ["can_eat"]
    }
  }, ["can_eat", "can_scoop_ice_cream"], "scope");
  expect(result).toBe(true);
});

test("returns true if in role in both scopes", () => {
  const result = hasPermission({
    roles: {
      __global_roles__: ["can_fry_bacon", "can_scoop_ice_cream"], // eslint-disable-line camelcase
      scope: ["can_eat", "can_scoop_ice_cream"]
    }
  }, ["can_scoop_ice_cream"], "scope");
  expect(result).toBe(true);
});

test("returns false if not in any role in either scope", () => {
  const result = hasPermission({
    roles: {
      __global_roles__: ["can_fry_bacon", "can_scoop_ice_cream"], // eslint-disable-line camelcase
      scope: ["can_eat", "can_scoop_ice_cream"]
    }
  }, ["can_do_dishes"], "scope");
  expect(result).toBe(false);
});

test("returns true if has owner role, even if not explicitly in the permissions array", () => {
  const result = hasPermission({
    roles: {
      __global_roles__: ["can_fry_bacon", "can_scoop_ice_cream"], // eslint-disable-line camelcase
      scope: ["can_eat", "can_scoop_ice_cream", "owner"]
    }
  }, ["can_do_dishes"], "scope");
  expect(result).toBe(true);
});
