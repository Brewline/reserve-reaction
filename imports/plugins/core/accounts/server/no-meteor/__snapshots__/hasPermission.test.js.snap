// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if permissions is not an array 1`] = `"permissions must be an array of strings"`;

exports[`throws if roleGroup is present but an empty string 1`] = `"roleGroup must be a non-empty string"`;

exports[`throws if roleGroup is present but not a string 1`] = `"roleGroup must be a non-empty string"`;
